File: backend/src/main/java/com/davnig/invoicemanagementapi/model/dto/SortConverter.java
package com.davnig.invoicemanagementapi.model.dto;


import lombok.NonNull;
import org.springframework.data.domain.Sort;

public class SortConverter {

    public enum SortDirection {
        ASCENDING("+") {
            @Override
            public Sort.Direction parse(String value) {
                return Sort.Direction.ASC;
            }
        },
        DESCENDING("-") {
            @Override
            public Sort.Direction parse(String value) {
                return Sort.Direction.DESC;
            }
        };

        private final String value;

        SortDirection(String value) {
            this.value = value;
        }

        public abstract Sort.Direction parse(String value);
    }

    /**
     * Parse the given {@literal String} in a {@link Sort} object.
     *
     * @param sortOrder a field name prefixed with "+" (ascending) or "-" (descending), e.g. "+id"
     * @return {@link Sort}
     */
    public static Sort parse(@NonNull String sortOrder) {
        if (sortOrder.length() < 1) throw new IllegalArgumentException();

        String direction = String.valueOf(sortOrder.charAt(0));
        for (SortDirection el : SortDirection.values()) {
            if (el.value.equals(direction)) {
                return Sort.by(el.parse(direction), sortOrder.substring(1));
            }
        }
        throw new IllegalArgumentException();
    }

}

File: backend/src/main/kotlin/com/davnig/invoicemanagementapi/model/dto/SortConverter.kt
package com.davnig.invoicemanagementapi.model.dto

import org.springframework.data.domain.Sort

class SortConverter {

    enum class SortDirection(val value: String) {

        ASCENDING("+") {
            override fun parse(): Sort.Direction {
                return Sort.Direction.ASC
            }
        },
        DESCENDING("-") {
            override fun parse(): Sort.Direction {
                return Sort.Direction.DESC
            }
        };

        abstract fun parse(): Sort.Direction

    }

    companion object {
        fun parse(sortOrder: String): Sort {
            return when (sortOrder[0].toString()) {
                SortDirection.ASCENDING.value -> Sort.by(Sort.Direction.ASC, sortOrder.substring(1))
                SortDirection.DESCENDING.value -> Sort.by(Sort.Direction.DESC, sortOrder.substring(1))
                else -> {
                    throw IllegalArgumentException()
                }
            }
        }
    }

}